src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/CodeActions/Razor/CodeBlockService.cs

```diff
@@ -0,0 +1,114 @@
+        var insertPosition = new Position(insertLineLocation.LineIndex, openBraceLocation.LineIndex == codeBlockEndBraceLocation.LineIndex ? codeBlockEndBraceLocation.CharacterIndex - 1 : 0);
+        return new TextEdit() { Range = new Range { Start = insertPosition, End = insertPosition }, NewText = formattedGeneratedMethod.Replace("2", methodName) };
+    }
+
+    private static string FormatMethodInCodeBlock(RazorSourceDocument source, SourceLocation codeBlockStartLocation, SourceLocation openBraceLocation, SourceLocation codeBlockEndLocation, SourceLocation insertLocation, int indentSize, string method)
+    {
+        var formattedGeneratedMethod = GenerateMethodIndentService.AddIndentation(method, indentSize, codeBlockStartLocation.CharacterIndex);
+        if (openBraceLocation.LineIndex == codeBlockEndLocation.LineIndex)
+        {
+            // The @code block's '{' and '}' are on the same line, we'll need to add a new line to both the beginning and end of the generated code.
+            return $"\r\n{formattedGeneratedMethod}\r\n";
+        }
+
+        if (insertLocation.LineIndex == codeBlockEndLocation.LineIndex)
+        {
+            // We will be inserting the code on the same line as the '}' of the code block. Make sure to add a new line to separate these.
+            formattedGeneratedMethod += "\r\n";
+        }
+
+        if (insertLocation.LineIndex - 1 == openBraceLocation.LineIndex)
+        {
+            // There is no other code in the @code block, no need to continue formatting.
+            return formattedGeneratedMethod;
+        }
+
+        // There is other code that exists in the @code block. Look at what is above the line we are going to insert at.
+        // If there is code above it, we need to add a new line at the beginning the generated code.
+        var previousLine = source.Lines.GetLocation(insertLocation.AbsoluteIndex - insertLocation.CharacterIndex - 1);
+        if (!IsLineEmpty(source, previousLine))
+        {
+            formattedGeneratedMethod = $"\r\n{formattedGeneratedMethod}";
+        }
+
+        return formattedGeneratedMethod;
+    }
+
+    /// <summary>
+    ///  Determines whether the line is empty.
+    /// </summary>
+    /// <param name="source">The RazorSourceDocument of where the line in question lives.</param>
+    /// <param name="endLineLocation">
+    ///  The <see cref="CodeAnalysis.SourceLocation"/> of the end of the line in question.
+    /// </param>
+    /// <returns>true if the line is empty, otherwise false.</returns>
+    private static bool IsLineEmpty(RazorSourceDocument source, SourceLocation endLineLocation)
+    {
+        var beginningLineIndex = endLineLocation.AbsoluteIndex - endLineLocation.CharacterIndex;
+        for(var i = endLineLocation.AbsoluteIndex; i >= beginningLineIndex; i--)
+        {
+            if (source[i] != '\r' && source[i] != '\n' && source[i] != ' ')
```

Comment:
Perhaps `char.IsWhitespace`? Though I think there might be a better way to do this, but will come back to it later.

Reply:
I took a look and this helper is only available for `SourceText`, `string` , and `TextLine`. It seems that it would be more work to get from what we have here to any of those types to use that helper than to just do it ourselves through this method.